File: src/python/universal_axiom.py
"""
The Universal Axiom - Core Implementation

Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)

Foundation Layer: Impulses (A), Elements (B), Pressure (C)
Dynamic Layer: Exponential Growth (E_n), Fibonacci Sequence (F_n)
Cognitive Layer: Subjectivity Scale (X), Why Axis (Y), TimeSphere (Z)
"""

from typing import Dict, List
from dataclasses import dataclass
import sys

# Maximum allowed value for n to prevent overflow
# At n=70, Fibonacci exceeds typical integer limits
# At n=33, E_n exceeds float precision
# We use 100 as a practical limit
MAX_N = 100

# Maximum safe value to prevent overflow
MAX_SAFE_VALUE = sys.float_info.max / 2


@dataclass
class FoundationLayer:
    """Foundation Layer: A · B · C"""

    impulses: float  # A - Fundamental drives
    elements: float  # B - Core components
    pressure: float  # C - Constraints and forces

    def compute(self) -> float:
        """Compute foundation layer product"""
        return self.impulses * self.elements * self.pressure


class DynamicLayer:
    """Dynamic Layer: E_n · (1 + F_n)"""

    def __init__(self, n: int = 1, base_exponential: float = 3.0):
        """
        Initialize DynamicLayer with validated n.

        Args:
            n: Current iteration/step (will be clamped to [1, MAX_N])
            base_exponential: Base for exponential growth
        """
        self._n = max(1, min(MAX_N, int(n)))
        self.base_exponential = base_exponential

    @property
    def n(self) -> int:
        """Get current iteration step"""
        return self._n

    @n.setter
    def n(self, value: int) -> None:
        """Set iteration step with validation"""
        self._n = max(1, min(MAX_N, int(value)))

    def exponential_growth(self) -> float:
        """E_n - Exponential growth component with overflow protection"""
        result = (2 * (self.base_exponential ** self._n)) - 1
        # Return MAX_SAFE_VALUE if overflow would occur
        if result > MAX_SAFE_VALUE or result != result:  # NaN check
            return MAX_SAFE_VALUE
        return result

    def fibonacci(self) -> int:
        """F_n - Fibonacci sequence for natural regulation with overflow protection"""
        if self._n <= 1:
            return 1

        a, b = 1, 1
        for _ in range(2, self._n + 1):
            next_val = a + b
            # Check for overflow
            if next_val > MAX_SAFE_VALUE:
                return int(MAX_SAFE_VALUE)
            a, b = b, next_val
        return b

    def compute(self) -> float:
        """Compute dynamic layer: E_n · (1 + F_n) with overflow protection"""
        E_n = self.exponential_growth()
        F_n = self.fibonacci()
        result = E_n * (1 + F_n)
        # Return MAX_SAFE_VALUE if overflow would occur
        if result > MAX_SAFE_VALUE or result != result:  # NaN check
            return MAX_SAFE_VALUE
        return result


@dataclass
class CognitiveLayer:
    """Cognitive Layer: X · Y · Z"""

    subjectivity: (
        float  # X - Objectivity measure (0 = fully objective, 1 = fully subjective)
    )
    purpose: float  # Y - Purpose-driven reasoning strength
    time: float  # Z - Temporal continuity and irreversibility

    def compute(self) -> float:
        """Compute cognitive layer product"""
        # X represents objectivity, so we invert: (1 - subjectivity) for objectivity score
        objectivity = 1 - self.subjectivity
        return objectivity * self.purpose * self.time


class UniversalAxiom:
    """
    The Universal Axiom - Complete Intelligence Model

    Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
    """

    def __init__(
        self,
        impulses: float = 1.0,
        elements: float = 1.0,
        pressure: float = 1.0,
        subjectivity: float = 0.0,
        purpose: float = 1.0,
        time: float = 1.0,
        n: int = 1,
    ):
        """
        Initialize The Universal Axiom

        Args:
            impulses: A - Fundamental drives (default: 1.0)
            elements: B - Core components (default: 1.0)
            pressure: C - Constraints and forces (default: 1.0)
            subjectivity: X - Subjectivity level 0-1 (default: 0.0 = fully objective)
            purpose: Y - Purpose strength (default: 1.0)
            time: Z - Temporal factor (default: 1.0)
            n: Current iteration step (default: 1)
        """
        self.foundation = FoundationLayer(impulses, elements, pressure)
        self.dynamic = DynamicLayer(n=n)
        self.cognitive = CognitiveLayer(subjectivity, purpose, time)
        self.n = max(1, min(MAX_N, int(n)))

    def compute_intelligence(self) -> float:
        """
        Compute Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)

        Returns:
            float: The computed intelligence value
        """
        foundation_value = self.foundation.compute()
        dynamic_value = self.dynamic.compute()
        cognitive_value = self.cognitive.compute()

        intelligence = dynamic_value * cognitive_value * foundation_value
        return intelligence

    def evolve(self, delta_time: float = 1.0) -> float:
        """
        Evolve the system forward in time
        Note: n is clamped to MAX_N to prevent overflow

        Args:
            delta_time: Time step increment

        Returns:
            float: New intelligence value after evolution
        """
        self.n = min(self.n + 1, MAX_N)
        self.dynamic.n = self.n
        self.cognitive.time += delta_time

        return self.compute_intelligence()

    def apply_pressure(self, pressure_delta: float) -> float:
        """
        Apply pressure change (e.g., from contradictions or constraints)

        Args:
            pressure_delta: Change in pressure (can be positive or negative)

        Returns:
            float: New intelligence value after pressure application
        """
        self.foundation.pressure += pressure_delta
        # Ensure pressure stays positive
        self.foundation.pressure = max(0.01, self.foundation.pressure)

        return self.compute_intelligence()

    def adjust_subjectivity(self, subjectivity_delta: float) -> float:
        """
        Adjust subjectivity level (moving toward objectivity or away)

        Args:
            subjectivity_delta: Change in subjectivity (-1 to +1)

        Returns:
            float: New intelligence value after adjustment
        """
        self.cognitive.subjectivity += subjectivity_delta
        # Clamp between 0 and 1
        self.cognitive.subjectivity = max(0.0, min(1.0, self.cognitive.subjectivity))

        return self.compute_intelligence()

    def strengthen_purpose(self, purpose_multiplier: float) -> float:
        """
        Strengthen or weaken purpose alignment

        Args:
            purpose_multiplier: Multiplier for purpose (e.g., 1.1 = 10% increase)

        Returns:
            float: New intelligence value after purpose adjustment
        """
        self.cognitive.purpose *= purpose_multiplier
        self.cognitive.purpose = max(0.01, self.cognitive.purpose)

        return self.compute_intelligence()

    def get_state(self) -> Dict:
        """
        Get current state of all variables

        Returns:
            Dict: Complete state dictionary
        """
        return {
            "n": self.n,
            "foundation": {
                "A_impulses": self.foundation.impulses,
                "B_elements": self.foundation.elements,
                "C_pressure": self.foundation.pressure,
                "product": self.foundation.compute(),
            },
            "dynamic": {
                "E_n": self.dynamic.exponential_growth(),
                "F_n": self.dynamic.fibonacci(),
                "product": self.dynamic.compute(),
            },
            "cognitive": {
                "X_subjectivity": self.cognitive.subjectivity,
                "X_objectivity": 1 - self.cognitive.subjectivity,
                "Y_purpose": self.cognitive.purpose,
                "Z_time": self.cognitive.time,
                "product": self.cognitive.compute(),
            },
            "intelligence": self.compute_intelligence(),
        }

    def __repr__(self) -> str:
        state = self.get_state()
        return f"UniversalAxiom(n={self.n}, Intelligence={state['intelligence']:.4f})"


class AxiomSimulator:
    """Simulator for running Universal Axiom scenarios"""

    def __init__(self, axiom: UniversalAxiom):
        self.axiom = axiom
        self.history: List[Dict] = []

    def record_state(self):
        """Record current state to history"""
        self.history.append(self.axiom.get_state())

    def simulate_evolution(
        self, steps: int = 10, delta_time: float = 1.0
    ) -> List[Dict]:
        """
        Simulate evolution over multiple time steps

        Args:
            steps: Number of evolution steps
            delta_time: Time increment per step

        Returns:
            List[Dict]: History of states
        """
        self.history = []
        self.record_state()

        for _ in range(steps):
            self.axiom.evolve(delta_time)
            self.record_state()

        return self.history

    def simulate_contradiction_resolution(
        self, initial_pressure: float = 2.0, resolution_steps: int = 5
    ) -> List[Dict]:
        """
        Simulate how the system handles contradiction

        Args:
            initial_pressure: Initial pressure spike from contradiction
            resolution_steps: Steps to resolve the contradiction

        Returns:
            List[Dict]: History showing pressure resolution
        """
        self.history = []
        self.record_state()

        # Apply initial pressure spike
        self.axiom.apply_pressure(initial_pressure)
        self.record_state()

        # Gradually resolve through objectivity adjustment and pressure release
        for i in range(resolution_steps):
            # Reduce subjectivity (increase objectivity)
            self.axiom.adjust_subjectivity(-0.1)

            # Release pressure as understanding increases
            pressure_release = -initial_pressure / resolution_steps
            self.axiom.apply_pressure(pressure_release)

            # Evolve forward
            self.axiom.evolve()
            self.record_state()

        return self.history

    def get_coherence_metric(self) -> float:
        """
        Calculate coherence metric based on balance of components

        Returns:
            float: Coherence score (higher is more coherent)
        """
        state = self.axiom.get_state()

        # Coherence is high when:
        # - Objectivity is high (low subjectivity)
        # - Purpose is strong
        # - Pressure is moderate (not too high or low)
        # - Components are balanced

        objectivity_score = state["cognitive"]["X_objectivity"]
        purpose_score = min(state["cognitive"]["Y_purpose"] / 2.0, 1.0)
        pressure_score = 1.0 / (1.0 + abs(state["foundation"]["C_pressure"] - 1.0))

        coherence = (objectivity_score + purpose_score + pressure_score) / 3.0
        return coherence


def fibonacci_sequence(n: int) -> List[int]:
    """Generate Fibonacci sequence up to n terms"""
    if n <= 0:
        return []
    elif n == 1:
        return [1]

    sequence = [1, 1]
    for i in range(2, n):
        sequence.append(sequence[i - 1] + sequence[i - 2])
    return sequence

File: mcp-server/lib/universal-axiom.ts
/**
 * The Universal Axiom - Core Implementation (TypeScript)
 *
 * Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
 *
 * Foundation Layer: Impulses (A), Elements (B), Pressure (C)
 * Dynamic Layer: Exponential Growth (E_n), Fibonacci Sequence (F_n)
 * Cognitive Layer: Subjectivity Scale (X), Why Axis (Y), TimeSphere (Z)
 */

/**
 * Maximum allowed value for n to prevent overflow
 * At n=70, Fibonacci exceeds Number.MAX_SAFE_INTEGER
 * At n=33, E_n exceeds Number.MAX_SAFE_INTEGER
 * We use 100 as a practical limit
 */
export const MAX_N = 100;

/**
 * State interface for the Axiom system
 */
export interface AxiomState {
  n: number;
  foundation: {
    A_impulses: number;
    B_elements: number;
    C_pressure: number;
    product: number;
  };
  dynamic: {
    E_n: number;
    F_n: number;
    product: number;
  };
  cognitive: {
    X_subjectivity: number;
    X_objectivity: number;
    Y_purpose: number;
    Z_time: number;
    product: number;
  };
  intelligence: number;
}

/**
 * Foundation Layer: A · B · C
 */
export class FoundationLayer {
  constructor(
    public impulses: number,
    public elements: number,
    public pressure: number
  ) {}

  compute(): number {
    return this.impulses * this.elements * this.pressure;
  }
}

/**
 * Dynamic Layer: E_n · (1 + F_n)
 */
export class DynamicLayer {
  private _n: number;

  constructor(n: number, public baseExponential: number = 3) {
    // Validate and clamp n to valid range [1, MAX_N]
    this._n = Math.max(1, Math.min(MAX_N, Math.floor(n)));
  }

  get n(): number {
    return this._n;
  }

  set n(value: number) {
    // Validate and clamp n to valid range [1, MAX_N]
    this._n = Math.max(1, Math.min(MAX_N, Math.floor(value)));
  }

  exponentialGrowth(): number {
    const result = 2 * Math.pow(this.baseExponential, this._n) - 1;
    // Return MAX_SAFE_INTEGER if overflow would occur
    if (!Number.isFinite(result)) {
      return Number.MAX_SAFE_INTEGER;
    }
    return result;
  }

  fibonacci(): number {
    if (this._n <= 1) return 1;
    let a = 1;
    let b = 1;
    for (let i = 2; i <= this._n; i++) {
      const next = a + b;
      // Check for overflow
      if (!Number.isFinite(next) || next > Number.MAX_SAFE_INTEGER) {
        return Number.MAX_SAFE_INTEGER;
      }
      [a, b] = [b, next];
    }
    return b;
  }

  compute(): number {
    const E_n = this.exponentialGrowth();
    const F_n = this.fibonacci();
    const result = E_n * (1 + F_n);
    // Return MAX_SAFE_INTEGER if overflow would occur
    if (!Number.isFinite(result)) {
      return Number.MAX_SAFE_INTEGER;
    }
    return result;
  }
}

/**
 * Cognitive Layer: X · Y · Z
 */
export class CognitiveLayer {
  constructor(
    public subjectivity: number, // 0-1 scale
    public purpose: number,
    public time: number
  ) {}

  compute(): number {
    // X represents objectivity: (1 - subjectivity)
    const objectivity = 1 - this.subjectivity;
    return objectivity * this.purpose * this.time;
  }
}

/**
 * Configuration for creating a UniversalAxiom instance
 */
export interface AxiomConfig {
  impulses?: number;
  elements?: number;
  pressure?: number;
  subjectivity?: number;
  purpose?: number;
  time?: number;
  n?: number;
}

/**
 * The Universal Axiom - Complete Intelligence Model
 *
 * Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
 */
export class UniversalAxiom {
  foundation: FoundationLayer;
  dynamic: DynamicLayer;
  cognitive: CognitiveLayer;
  n: number;

  constructor({
    impulses = 1.0,
    elements = 1.0,
    pressure = 1.0,
    subjectivity = 0.0,
    purpose = 1.0,
    time = 1.0,
    n = 1,
  }: AxiomConfig = {}) {
    this.foundation = new FoundationLayer(impulses, elements, pressure);
    this.dynamic = new DynamicLayer(n);
    this.cognitive = new CognitiveLayer(subjectivity, purpose, time);
    this.n = n;
  }

  /**
   * Compute Intelligence_n = E_n · (1 + F_n) · X · Y · Z · (A · B · C)
   */
  computeIntelligence(): number {
    const foundationValue = this.foundation.compute();
    const dynamicValue = this.dynamic.compute();
    const cognitiveValue = this.cognitive.compute();
    return dynamicValue * cognitiveValue * foundationValue;
  }

  /**
   * Evolve the system forward in time
   * Note: n is clamped to MAX_N to prevent overflow
   */
  evolve(deltaTime: number = 1.0): number {
    this.n = Math.min(this.n + 1, MAX_N);
    this.dynamic.n = this.n;
    this.cognitive.time += deltaTime;
    return this.computeIntelligence();
  }

  /**
   * Apply pressure change (e.g., from contradictions or constraints)
   */
  applyPressure(pressureDelta: number): number {
    this.foundation.pressure += pressureDelta;
    // Ensure pressure stays positive
    this.foundation.pressure = Math.max(0.01, this.foundation.pressure);
    return this.computeIntelligence();
  }

  /**
   * Adjust subjectivity level (moving toward objectivity or away)
   */
  adjustSubjectivity(subjectivityDelta: number): number {
    this.cognitive.subjectivity += subjectivityDelta;
    // Clamp between 0 and 1
    this.cognitive.subjectivity = Math.max(
      0.0,
      Math.min(1.0, this.cognitive.subjectivity)
    );
    return this.computeIntelligence();
  }

  /**
   * Strengthen or weaken purpose alignment
   */
  strengthenPurpose(purposeMultiplier: number): number {
    this.cognitive.purpose *= purposeMultiplier;
    this.cognitive.purpose = Math.max(0.01, this.cognitive.purpose);
    return this.computeIntelligence();
  }

  /**
   * Get current state of all variables
   */
  getState(): AxiomState {
    return {
      n: this.n,
      foundation: {
        A_impulses: this.foundation.impulses,
        B_elements: this.foundation.elements,
        C_pressure: this.foundation.pressure,
        product: this.foundation.compute(),
      },
      dynamic: {
        E_n: this.dynamic.exponentialGrowth(),
        F_n: this.dynamic.fibonacci(),
        product: this.dynamic.compute(),
      },
      cognitive: {
        X_subjectivity: this.cognitive.subjectivity,
        X_objectivity: 1 - this.cognitive.subjectivity,
        Y_purpose: this.cognitive.purpose,
        Z_time: this.cognitive.time,
        product: this.cognitive.compute(),
      },
      intelligence: this.computeIntelligence(),
    };
  }

  toString(): string {
    const state = this.getState();
    return `UniversalAxiom(n=${this.n}, Intelligence=${state.intelligence.toFixed(4)})`;
  }
}

/**
 * Simulator for running Universal Axiom scenarios
 */
export class AxiomSimulator {
  private history: AxiomState[] = [];

  constructor(private axiom: UniversalAxiom) {}

  private recordState(): void {
    this.history.push(this.axiom.getState());
  }

  /**
   * Simulate evolution over multiple time steps
   */
  simulateEvolution(steps: number = 10, deltaTime: number = 1.0): AxiomState[] {
    this.history = [];
    this.recordState();

    for (let i = 0; i < steps; i++) {
      this.axiom.evolve(deltaTime);
      this.recordState();
    }

    return this.history;
  }

  /**
   * Simulate how the system handles contradiction
   */
  simulateContradictionResolution(
    initialPressure: number = 2.0,
    resolutionSteps: number = 5
  ): AxiomState[] {
    this.history = [];
    this.recordState();

    // Apply initial pressure spike
    this.axiom.applyPressure(initialPressure);
    this.recordState();

    // Gradually resolve through objectivity adjustment and pressure release
    for (let i = 0; i < resolutionSteps; i++) {
      // Reduce subjectivity (increase objectivity)
      this.axiom.adjustSubjectivity(-0.1);

      // Release pressure as understanding increases
      const pressureRelease = -initialPressure / resolutionSteps;
      this.axiom.applyPressure(pressureRelease);

      // Evolve forward
      this.axiom.evolve();
      this.recordState();
    }

    return this.history;
  }

  /**
   * Calculate coherence metric based on balance of components
   */
  getCoherenceMetric(): number {
    const state = this.axiom.getState();

    // Coherence is high when:
    // - Objectivity is high (low subjectivity)
    // - Purpose is strong
    // - Pressure is moderate (not too high or low)
    const objectivityScore = state.cognitive.X_objectivity;
    const purposeScore = Math.min(state.cognitive.Y_purpose / 2.0, 1.0);
    const pressureScore =
      1.0 / (1.0 + Math.abs(state.foundation.C_pressure - 1.0));

    return (objectivityScore + purposeScore + pressureScore) / 3.0;
  }

  getHistory(): AxiomState[] {
    return this.history;
  }
}

/**
 * Generate Fibonacci sequence up to n terms
 */
export function fibonacciSequence(n: number): number[] {
  if (n <= 0) return [];
  if (n === 1) return [1];

  const sequence = [1, 1];
  for (let i = 2; i < n; i++) {
    sequence.push(sequence[i - 1] + sequence[i - 2]);
  }
  return sequence;
}
